fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/FDBRecordStoreUniqueIndexTest.java

```diff
@@ -303,74 +338,458 @@ public void changeIndexAtFixedSubspaceKey() throws Exception {
+     * what would be more violations if the index were still unique.
+     * @param allowReadableUniquePending whether to allow {@link IndexState#READABLE_UNIQUE_PENDING}
+     * @throws Exception if there is an issue
+     */
+    @ParameterizedTest(name = "removeUniquenessConstraintDuringTransactionWithNewDuplications(addViolations={0}, allowReadableUniquePending={1})")
+    @MethodSource("removeUniquenessConstraintArguments")
+    void removeUniquenessConstraintDuringTransactionWithNewDuplications(boolean allowReadableUniquePending) throws Exception {
+        final DropUniquenessConstraint dropUniquenessConstraint = new DropUniquenessConstraint(allowReadableUniquePending);
+        dropUniquenessConstraint.setupStore();
+        dropUniquenessConstraint.addUniqueIndexViaBuild();
+        dropUniquenessConstraint.changeToNonUnique(false, false, true);
+    }
+
+    /**
+     * This test covers the situation where you add some violations, and then, in the same transaction, change the index
+     * to no longer be unique.
+     * <p>
+     *     Right now this will fail {@link FDBRecordStore#checkVersion(FDBRecordStoreBase.UserVersionChecker, FDBRecordStoreBase.StoreExistenceCheck)}
+     *     and fail the commit. In theory, we could support this clearing out the index violation checks, but that
+     *     proved tricky, and this seems like pathological use case.
+     * </p>
+     * @param allowReadableUniquePending whether to allow {@link IndexState#READABLE_UNIQUE_PENDING}
+     * @throws Exception if there is an issue
+     */
+    @ParameterizedTest(name = "removeUniquenessConstraintDuringTransactionWithNewViolations(allowReadableUniquePending={0})")
+    @BooleanSource
+    void removeUniquenessConstraintDuringTransactionWithNewViolations(boolean allowReadableUniquePending) throws Exception {
+        final DropUniquenessConstraint dropUniquenessConstraint = new DropUniquenessConstraint(allowReadableUniquePending);
+        dropUniquenessConstraint.setupStore();
+        dropUniquenessConstraint.addUniqueIndexViaBuild();
+        dropUniquenessConstraint.changeToNonUnique(false, true, false);
+    }
+
+    /**
+     * This test is primarily here to make sure that if there is an index maintainer that doesn't implement the new
+     * {@link IndexMaintainer#clearUniquenessViolations}, the old flow of increasing the version when changing the state
+     * still works.
+     * @param allowReadableUniquePending whether to allow {@link IndexState#READABLE_UNIQUE_PENDING}
+     * @throws Exception if there is an issue
+     */
+    @ParameterizedTest(name = "bumpVersionWhenChangingToNonUnique(allowReadableUniquePending={0})")
+    @BooleanSource
+    void bumpVersionWhenChangingToNonUnique(boolean allowReadableUniquePending) throws Exception {
+        final DropUniquenessConstraint dropUniquenessConstraint = new DropUniquenessConstraint(
+                allowReadableUniquePending, NO_UNIQUE_CLEAR_INDEX_TYPE, true);
```

Comment:
Is it worth adding a test of what happens with the `NO_UNIQUE_CLEAR_INDEX_TYPE` type and we don't bump the version? I think I'd expect it to act like the index doesn't have any uniqueness constraints, but we'd still be able to find some by scanning the index violations subspace